packages/pigeon/platform_tests/test_plugin/android/src/main/kotlin/com/example/test_plugin/ProxyApiTests.gen.kt

```diff
@@ -0,0 +1,475 @@
+                wrapError(exception)
+              }
+          reply.reply(wrapped)
+        }
+      }
+    }
+  }
+
+  fun removeStrongReference(identifier: Long, callback: (Result<Unit>) -> Unit) {
+    val channelName =
+        "dev.flutter.pigeon.pigeon_integration_tests.PigeonInstanceManagerApi.removeStrongReference"
+    val channel = BasicMessageChannel<Any?>(binaryMessenger, channelName, codec)
+    channel.send(identifier) {
+      if (it is List<*>) {
+        if (it.size > 1) {
+          callback(
+              Result.failure(
+                  ProxyApiTestsError(it[0] as String, it[1] as String, it[2] as String?)))
+        } else {
+          callback(Result.success(Unit))
+        }
+      } else {
+        callback(Result.failure(createConnectionError(channelName)))
+      }
+    }
+  }
+}
+
+abstract class PigeonProxyApiBaseCodec(
+    val binaryMessenger: BinaryMessenger,
+    val instanceManager: PigeonInstanceManager
+) : StandardMessageCodec() {
+  /**
+   * An implementation of [PigeonApiProxyApiTestClass] used to add a new Dart instance of
+   * `ProxyApiTestClass` to the Dart `InstanceManager`.
+   */
+  abstract fun getPigeonApiProxyApiTestClass(): PigeonApiProxyApiTestClass
+
+  /**
+   * An implementation of [PigeonApiProxyApiSuperClass] used to add a new Dart instance of
+   * `ProxyApiSuperClass` to the Dart `InstanceManager`.
+   */
+  abstract fun getPigeonApiProxyApiSuperClass(): PigeonApiProxyApiSuperClass
+
+  /**
+   * An implementation of [PigeonApiProxyApiInterface] used to add a new Dart instance of
+   * `ProxyApiInterface` to the Dart `InstanceManager`.
+   */
+  abstract fun getPigeonApiProxyApiInterface(): PigeonApiProxyApiInterface
+
+  fun setUpMessageHandlers() {
+    PigeonApiProxyApiTestClass.setUpMessageHandlers(
+        binaryMessenger, getPigeonApiProxyApiTestClass())
+    PigeonApiProxyApiSuperClass.setUpMessageHandlers(
+        binaryMessenger, getPigeonApiProxyApiSuperClass())
+  }
+
+  override fun readValueOfType(type: Byte, buffer: ByteBuffer): Any? {
+    return when (type) {
+      128.toByte() -> {
+        return instanceManager.getInstance(
+            readValue(buffer).let { if (it is Int) it.toLong() else it as Long })
+      }
+      else -> super.readValueOfType(type, buffer)
+    }
+  }
+
+  override fun writeValue(stream: ByteArrayOutputStream, value: Any?) {
+    if (value is ProxyApiTestClass) {
+      getPigeonApiProxyApiTestClass().pigeon_newInstance(value) {}
+    } else if (value is com.example.test_plugin.ProxyApiSuperClass) {
+      getPigeonApiProxyApiSuperClass().pigeon_newInstance(value) {}
+    } else if (value is ProxyApiInterface) {
+      getPigeonApiProxyApiInterface().pigeon_newInstance(value) {}
+    }
+
+    when {
+      instanceManager.containsInstance(value) -> {
+        stream.write(128)
+        writeValue(stream, instanceManager.getIdentifierForStrongReference(value))
+      }
+      else -> super.writeValue(stream, value)
+    }
+  }
+}
```

Comment:
@camsim99 FYI

@stuartmorgan @tarrinneal This is one of the places where the design changes from the implementation in the doc.  I wanted to improve the scenario of when a method returns a `ProxyApi` or a list/map of a `ProxyApi`:

```java
class Basket {
  Apple giveMeAnApple();
  List<Apple> giveMeApples();
}
```

Typically we solve this by needing to handle this manually when we implement the api:

```java
class  PigeonApiBasketImpl {
  Apple giveMeAnApple(Basket instance) {
    final Apple apple = instance.giveMeAnApple();
    if (!instanceManager.containsInstance(apple)) {
      instanceManager.addHostCreatedInstance(apple);
    }

    return apple;
  }

  List<Apple> giveMeApples(Basket instance) {
    final List<Apple> apples = instance.giveMeApples();
    for (Apple apple in apples) {
      if (!instanceManager.containsInstance(apple)) {
        instanceManager.addHostCreatedInstance(apple);
      }
    }

    return apples;
  }
}
```

This change makes the codec the central class that contains all the proxy api implementations, so the `writeValue` can handle creating new dart instances by calling `pigeon_newInstance`.

This also:
* Creates a registrar like class that ensures a user implements each proxy api.
* A central class with every api implementation prevents the need for multiple instances of the same api implementation.

Reply:
This may end up overlapping with some of the codec work I'm about to do as well. I think it's generally a good idea to create a central codec though. We may end up with multiple codecs implementing each other.

Reply:
My main concern at this point is that the number of custom inputs is limited to around 120, so if there are a lot of proxy apis, enums, and or data classes, we may not be able to accommodate them all.

Reply:
As of right now, the generation of the `ProxyApi` codec is separate from Host/Flutter APIs codecs. Mainly because:
1. ProxyApis don't support data classes
2. ProxyApis support other ProxyApis in the codec and Host/Flutter APIs don't.

So they currently don't really share anything. If you create a central custom codec for all the APIs, then `ProxyApis` should only need one dedicated spot that uses the `InstanceManager`: https://github.com/flutter/packages/blob/main/packages/pigeon/platform_tests/shared_test_plugin_code/lib/src/generated/proxy_api_tests.gen.dart#L374. 

Also, as a side note, since all data classes have an `encode/decode` could you create a base data class for all the data classes and only use one spot.